### PR TITLE
Add unit test coverage for HcaManage

### PR DESCRIPTION
**Pull request by ThePletch**
## Why

[Relevant ticket](https://broadinstitute.atlassian.net/browse/DSPDC-1570)

Our HCA Manage code is untested, which makes it difficult to modify safely (and will make it difficult to refactor later).

## This PR

* Adds unit tests for the internal logic in HcaManage, including its interactions with BigQuery and the data repo
* Adds some additional test utilities, e.g. helpful argument matcher classes and mocks for GCS objects
* Performs some minor cleanups (e.g. moving private methods, i.e. underscore-prefixed ones, to the end of the class definition, fixing a couple logging statements and indentations)
* DOES NOT directly test BigQuery queries or functions that assemble other functions (e.g. it tests _process_rows and the matchers/lookups that can feed into it, but does not test functions that pass specific matchers and lookups to _process_rows)

### CODE REVIEW

File: orchestration/dagster_orchestration/hca_manage/tests/test_hca_manage.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+    def test_get_dataset_id_gets_id_of_first_dataset(self):
+        enumerate_datasets_response = Mock()
+        first_dataset = Mock()
+        first_dataset.id = 'abc'
+
+        second_dataset = Mock()
+        second_dataset.id = 'def'
+        enumerate_datasets_response.items = [first_dataset, second_dataset]
+        self.manager.data_repo_client.enumerate_datasets.return_value = enumerate_datasets_response
```
Comment (aherbst-broad) [inline orchestration/dagster_orchestration/hca_manage/tests/test_hca_manage.py:109]: Might be out of scope: the manager should raise if there is more than one dataset with the given name.

File: orchestration/dagster_orchestration/hca_manage/tests/test_hca_manage.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    def test_submit_snapshot_request_uses_qualifier_if_present(self):
+        self.manager.submit_snapshot_request(qualifier='steve')
+        self.manager.data_repo_client.create_snapshot.assert_called_once_with(
+            snapshot=ObjectWithAttributes(
+                SnapshotRequestModel,
+                name=StringMatchingRegex(r'datasetname___\d{8}_steve'))
+        )
+
+    def test_delete_snapshot_fetches_id_if_missing(self):
+        enumerate_snapshots_response = Mock()
```
Comment (aherbst-broad) [inline orchestration/dagster_orchestration/hca_manage/tests/test_hca_manage.py:153]: General nit: for readability please try to structure your tests like the following:

```
<preamble/setup>

<invocation of unit under test>

<assertions>
```

so in this case:

```suggestion
        enumerate_snapshots_response = Mock()
        single_snapshot = Mock()
        single_snapshot.id = 'abc'
        enumerate_snapshots_response.items = [single_snapshot]
        self.manager.data_repo_client.enumerate_snapshots.return_value = enumerate_snapshots_response
        
        self.manager.delete_snapshot(snapshot_name='steve')
        
        self.manager.data_repo_client.delete_snapshot.assert_called_once_with('abc')
```

File: orchestration/dagster_orchestration/hca_manage/diff_dirs.py [hunk 3]
```diff
@@ -17,8 +16,8 @@ def diff_dirs(
     source_bucket: str,
     source_prefix: str,
     target_bucket: str,
-    target_prefix: str
-) -> tuple[dict[str, int], dict[str, int]]:
+    target_prefix: str,
+) -> tuple[dict[str, str], dict[str, str]]:
```
Comment (aherbst-broad) [inline orchestration/dagster_orchestration/hca_manage/diff_dirs.py:20]: 👍 good catch

File: orchestration/dagster_orchestration/hca_manage/manage.py [hunk 4]
```diff
@@ -182,18 +182,6 @@ def get_null_filerefs(self, target_table: str) -> set[str]:
 
         return self._hit_bigquery(query)
 
-    def _hit_bigquery(self, query: str) -> set[str]:
```
Comment (aherbst-broad) [inline orchestration/dagster_orchestration/hca_manage/manage.py:185]: ❓ I assume the premise here is to keep the private methods at the bottom of the file?

File: orchestration/dagster_orchestration/hca_manage/manage.py [hunk 5]
```diff
@@ -342,7 +305,7 @@ def process_dangling_proj_refs(self, soft_delete: bool = False) -> int:
         :return: Number of rows with dangling project refs
         """
         if soft_delete:
-            raise Exception("Soft deleting rows with dangling project refs is unsupported")
+            raise NotImplementedError("Soft deleting rows with dangling project refs is unsupported")
```
Comment (aherbst-broad) [inline orchestration/dagster_orchestration/hca_manage/manage.py:308]: 👍 more targeted

**Review by aherbst-broad — APPROVED** [review]
Looks good.

File: orchestration/dagster_orchestration/hca_manage/manage.py (hunk 4 above)
Comment (ThePletch) [reply orchestration/dagster_orchestration/hca_manage/manage.py:185]: that's the idea, yeah. i'm considering adding a comment divider.

my other thought was just that the delineation between which methods are public and which are private is a bit unclear, but that's a topic for the refactor

File: orchestration/dagster_orchestration/hca_manage/tests/test_hca_manage.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+            )
+        self.assertEqual(result, 2)  # one error per table
+
+    def test_soft_delete_rows_calls_submit_soft_delete(self):
+        fake_google_path = 'fake://google/path.csv'
+        fake_job_id = 'jorb_id'
+        patched_soft_delete = patch('hca_manage.manage.HcaManage._submit_soft_delete', return_value=fake_job_id)
+        patched_put_in_bucket = patch('hca_manage.manage.HcaManage.put_csv_in_bucket', return_value=fake_google_path)
+
+        with patched_soft_delete as mock_submit, patched_put_in_bucket as mock_bucket_upload:
```
Comment (raaidbroad) [inline orchestration/dagster_orchestration/hca_manage/tests/test_hca_manage.py:99]: Did not know that you can just comma-separate with-as statements whoa

**Review by raaidbroad — APPROVED** [review]
beautiful 💯

**Pull request merged**